swiperefresh/src/main/java/com/google/accompanist/swiperefresh/SwipeRefresh.kt

```diff
@@ -0,0 +1,336 @@
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.Stable
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberCoroutineScope
+import androidx.compose.runtime.rememberUpdatedState
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.draw.clipToBounds
+import androidx.compose.ui.geometry.Offset
+import androidx.compose.ui.input.nestedscroll.NestedScrollConnection
+import androidx.compose.ui.input.nestedscroll.NestedScrollSource
+import androidx.compose.ui.input.nestedscroll.nestedScroll
+import androidx.compose.ui.layout.Layout
+import androidx.compose.ui.layout.layoutId
+import androidx.compose.ui.unit.Velocity
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+import kotlin.math.absoluteValue
+import kotlin.math.max
+
+private const val DragMultiplier = 0.5f
+
+/**
+ * Creates a [SwipeRefreshState] that is remembered across compositions.
+ *
+ * Changes to [isRefreshing] will result in the [SwipeRefreshState] being updated.
+ *
+ * @param isRefreshing the value for [SwipeRefreshState.isRefreshing]
+ */
+@Composable
+fun rememberSwipeRefreshState(
+    isRefreshing: Boolean
+): SwipeRefreshState {
+    return remember {
+        SwipeRefreshState(
+            isRefreshing = isRefreshing
+        )
+    }.apply {
+        this.isRefreshing = isRefreshing
+    }
+}
+
+/**
+ * A state object that can be hoisted to control and observe changes for [SwipeRefresh].
+ *
+ * In most cases, this will be created via [rememberSwipeRefreshState].
+ *
+ * @param isRefreshing the initial value for [SwipeRefreshState.isRefreshing]
+ */
+@Stable
+class SwipeRefreshState(
+    isRefreshing: Boolean,
+) {
+    private val _indicatorOffset = Animatable(0f)
+    private val mutatorMutex = MutatorMutex()
+
+    internal var refreshTrigger by mutableStateOf(0f)
+
+    /**
+     * Whether this [SwipeRefreshState] is currently refreshing or not.
+     */
+    var isRefreshing: Boolean by mutableStateOf(isRefreshing)
+
+    /**
+     * Whether a swipe/drag is currently in progress.
+     */
+    var isSwipeInProgress: Boolean by mutableStateOf(false)
+        internal set
+
+    /**
+     * The current offset for the indicator, in pixels.
+     */
+    val indicatorOffset: Float get() = _indicatorOffset.value
+
+    internal suspend fun animateOffsetTo(offset: Float) {
+        mutatorMutex.mutate {
+            _indicatorOffset.animateTo(offset)
+        }
+    }
+
+    internal suspend fun animateBackToRest() {
+        mutatorMutex.mutate {
+            _indicatorOffset.animateTo(if (isRefreshing) refreshTrigger else 0f)
+        }
+    }
+
+    /**
+     * Dispatch scroll delta in pixels.
+     */
+    suspend fun dispatchRawDelta(delta: Float) {
+        mutatorMutex.mutate {
```

Comment:
I guess this method should either not lock the mutex at all (since it's a "raw" dispatch), or lock the mutex with the `MutationPriority.UserInput` since we're reacting on the user input while dispatching raw deltas.

Reply:
Went with `MutationPriority.UserInput`. Maybe I should rename this to `onDragDelta()` or something?

Reply:
yeah, might be worth it. It's not so raw anymore since you are acquiring mutex here :)

On a related note: do you really want it to be public?

Reply:
Done, and made `internal`